# LECTURE 13-02-2023

## VERY FIRST STEPS

Exercise 1

We can use a variable `msg` or not.

In [ ]:
msg = 'Hello World!'
print(msg)

Hello World!


In [ ]:
print('Hello World!')

Hello World!


# Types of variables

Variables can be of type:


* *numerical* (`integer` or `float`)
* *boolean* (`True` or `False`)
* *string* (text always between `''` or `""`)

To check the type of the variable we just need to use the function `type`.




Exercise 2

In [ ]:
x = 10
x

10

In [ ]:
type(x)

int

In [ ]:
x = 2.3
x

2.3

In [ ]:
type(x)

float

In [ ]:
w = "Pedro"
w

'Pedro'

In [ ]:
type(w)

str

In [ ]:
y = True

In [ ]:
type(y)

bool

## LISTS



1. Lists are collections of elements (numbers, strings, etc.)
2. The elements of a list can be accessed trhough indexing.
3. Let `1` be a list. To access the element at position `i`(index`i`), we just use: `1[i]`



In [ ]:
x = [2,8,9,10]
x

[2, 8, 9, 10]

How to get the first element of the list?

In [ ]:
x[0]

2

How to get the last element of a list?

In [ ]:
x[-1]

10

In [ ]:
x[3]

10

The element just before the last one has index `-2`, and so on.

How to get all the elements of a list after index `i`?

We just need to run `mylist[i:]`.
Note that, in this case, the element at index `i` is also returned.

In [ ]:
x[1:]

[8, 9, 10]

To get all the elements of a list from index `i` (inclusive) to index `j` (exclusive), we just need to run: `mylist[i:j]`

In [ ]:
x[1:3]

[8, 9]

Exercise 3,4 a) making explicit all numbers.

In [ ]:
l = [0,1,2,3,4,5,6,7,8,9,10]
l

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

  Exercise 3,4 b) using function `range`.

  Notice that `range(k)` does not include `k` in the range it creates. That's why when we want a list of numbers until 10, we must write 10+1 in order to 10 appears.


In [ ]:
l = list(range(10+1))
l

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

In [ ]:
m = list(range(210+1))
m

[0,
 1,
 2,
 3,
 4,
 5,
 6,
 7,
 8,
 9,
 10,
 11,
 12,
 13,
 14,
 15,
 16,
 17,
 18,
 19,
 20,
 21,
 22,
 23,
 24,
 25,
 26,
 27,
 28,
 29,
 30,
 31,
 32,
 33,
 34,
 35,
 36,
 37,
 38,
 39,
 40,
 41,
 42,
 43,
 44,
 45,
 46,
 47,
 48,
 49,
 50,
 51,
 52,
 53,
 54,
 55,
 56,
 57,
 58,
 59,
 60,
 61,
 62,
 63,
 64,
 65,
 66,
 67,
 68,
 69,
 70,
 71,
 72,
 73,
 74,
 75,
 76,
 77,
 78,
 79,
 80,
 81,
 82,
 83,
 84,
 85,
 86,
 87,
 88,
 89,
 90,
 91,
 92,
 93,
 94,
 95,
 96,
 97,
 98,
 99,
 100,
 101,
 102,
 103,
 104,
 105,
 106,
 107,
 108,
 109,
 110,
 111,
 112,
 113,
 114,
 115,
 116,
 117,
 118,
 119,
 120,
 121,
 122,
 123,
 124,
 125,
 126,
 127,
 128,
 129,
 130,
 131,
 132,
 133,
 134,
 135,
 136,
 137,
 138,
 139,
 140,
 141,
 142,
 143,
 144,
 145,
 146,
 147,
 148,
 149,
 150,
 151,
 152,
 153,
 154,
 155,
 156,
 157,
 158,
 159,
 160,
 161,
 162,
 163,
 164,
 165,
 166,
 167,
 168,
 169,
 170,
 171,
 172,
 173,
 174,
 175,
 176,
 177,
 178,
 179,
 180,
 181,
 182,
 183,
 184,


Exercise 3,4 c) confirm that you have really created a list.

In [ ]:
type(l)

list

In [ ]:
type(m)

list

# FOR LOOP

**What does a for loop do?**
A for loop is a control flow statement for specifying iteration, which allows code to be executed repeatedly. A for loop has two parts: a header specifying the iteration, and a body which is executed once per iteration.

Exercise 5

In [ ]:
names = ['Mary', 'John', 'Catherine']
marks = [9, 18, 19]

The "f" in the provided code is an abbreviation for "formatted string literal" or "f-string" in Python.

F-strings provide a concise and convenient way to embed expressions inside string literals, using curly braces {} to indicate where the expressions should be replaced with their values.

In [ ]:
print(f'The score of {names[0]} is {marks[0]}.')
print(f'The score of {names[1]} is {marks[1]}.')
print(f'The score of {names[2]} is {marks[2]}.')

The score of Mary is 9.
The score of John is 18.
The score of Catherine is 19.


* The `for` loop is central to all programming language, as it allows us to program repetitive operations using a very compact piece of code.
* In the following code cell, the variable `i` will successively assume the values in the specific range.
* Function `len` gives us the length of the list names.
* The length of the list names is `3` (it has 3 elements). Hence, `range(len(names))` is equivalent to `range(3)`.
* `range(3)` corresponds to the list `[0,1,2]`.
*As we can see, `i` will take, successively, the values of the range, i.e, 0,1,2.











In [ ]:
for i in range(len(names)):
  print(i)

0
1
2


In [ ]:
for i in range(len(names)):
  print(f'The score of {names[i]} is {marks[i]}.')

The score of Mary is 9.
The score of John is 18.
The score of Catherine is 19.


Exercise 6

This exercise needs the construction `if...else` that works as follows:

`if (logical condition):
  do this if the logical condition is True
else:
  do this if the logical condition is False`

In [ ]:
for i in range(len(names)):
  if (marks[i] < 10):
    greeting = 'Keep studying!'
  else:
    greeting = 'Congratulations!'

print(f'The score of {names[i]} is {marks[i]}. {greeting}')


The score of Catherine is 19. Congratulations!


Exercise 7

Program that gives us the name and respective weight (by the respective credits) average of the score of each student.

In [ ]:
names = ['Mary', 'John', 'Catherine']
course1 = [15, 18, 19]
course2 = [18, 19, 14]
course3 = [14, 16, 17]
credits = [6, 4.5, 3] # of the above courses, respectively

Code for each individual:

In [ ]:
w_avg_mary = round((course1[0]*credits[0] + course2[0]*credits[1] + course3[0]*credits[2]) / sum(credits), 2)
w_avg_mary

15.78

In [ ]:
w_avg_john = round((course1[1]*credits[0] + course2[1]*credits[1] + course3[1]*credits[2]) / sum(credits), 2)
w_avg_john

17.89

In [ ]:
w_avg_catherine = round((course1[2]*credits[0] + course2[2]*credits[1] + course3[2]*credits[2]) / sum(credits), 2)
w_avg_catherine

16.89

For all of them in just a code.

In [ ]:
for i in range(len(names)):
  w_avg = round((course1[i]*credits[0] + course2[i]*credits[1] + course2[i]*credits[2]) / sum(credits), 2)

print(f'The weighted average of {names[i]} is {w_avg}.')

The weighted average of Catherine is 16.22.


## LIST COMPREHENSION

Exercise 8 Create a list with the square roots of the numbers of a certain list.

In [ ]:
mylist = [4,3,1,5]
mylist

[4, 3, 1, 5]

In [ ]:
[x**2 for x in mylist]

[16, 9, 1, 25]

# DICTIONAIRES

* Dictionaires are useful, as they can store information that can be retrivied (recuperadas) by using a key.

* For instance, with disctionary `d`, defined as 

`d = {'a': 4, 'b': 2, 'c': 5}`

we can access the second element by using is key, `b`, in the following way:

`d['b]`

Exercise 9 Create a dictionary with the names as key and the marks as values of the following lists.


In [ ]:
names = ['Mary', 'John', 'Catherine']
marks = [9, 18, 19]

Exercise 9 a) manual enumeration

In [ ]:
d = {'Mary': 15, 'John': 18, 'Catherine': 19}
d

{'Mary': 15, 'John': 18, 'Catherine': 19}

In [ ]:
d['Catherine']

19

Exercise 9 b) **ZIP FUNCTION**

In Python, zip is a built-in function that is used to combine two or more iterables into a single iterable. It takes multiple iterables (such as lists, tuples, or sets) as input and returns an iterator of tuples where the i-th tuple contains the i-th element from each of the input iterables.

* Zip function is useful, as it allows us to blind lists together pairwise (ocultar listas em pares).


In [ ]:
l1 = ['a', 'b', 'c']
l2 = [1,4,2]

zip(l1, l2)

* We do not see the result of `zip` function.
* However, we can use list comprehension to see the contents of the result of the `zip` function.



In [ ]:
[x for x in zip(l1, l2)]

[('a', 1), ('b', 4), ('c', 2)]

As we can see, each element of the result of the `zip` function is a tuple, which contains the element of `l1` and its respective element of `l2`.

In [ ]:
[x for x in zip(names, marks)]

[('Mary', 9), ('John', 18), ('Catherine', 19)]

Exercise 9 c) using dictionary comprehension.

* Dictionary comprehension works similarly to the way list comprehension works.
*Since each element of the result of `zip` function is a tuple and tuple indexing is similar to list indexing,
** `x[0]` contains the element of `names`, and
** `x[1]` contains the element of `marks`, when

`x` is the element of the result of `zip(names, marks)`.

In [ ]:
{x[0]: x[1] for x in zip(names, marks)}

{'Mary': 9, 'John': 18, 'Catherine': 19}

Exercise 9 d) create a dictionary using functions zip and dict.

In [ ]:
d2 = dict(zip(names, marks))
d2

{'Mary': 9, 'John': 18, 'Catherine': 19}

# WHAT ARE TUPLES?

* Tuples are very similar to lists, but they are unchangeable. We can create them by using `()`.
* Indexing works similarly to indexing in lists.
 

In [ ]:
t = (4,7,9)
t

(4, 7, 9)

In [ ]:
t[2]

9

Tuples are unchangeable. Therefore, when we try to change them, Python returns an error.

In [ ]:
t[2] = 10

TypeError: ignored

While tuples cannot be changed, lists can.

In [ ]:
names2 = names
names2

['Mary', 'John', 'Catherine']

In [ ]:
names2[1] = 'Peter'
names2

['Mary', 'Peter', 'Catherine']

## LECTURE 20-02-2023

# NUMPY ARRAYS

A NumPy array is a multi-dimensional data structure used for storing and manipulating numerical data in Python. It is a fundamental component of the scientific Python ecosystem, and is used extensively in fields such as machine learning, data science, and scientific computing.

NumPy arrays are similar to Python lists, but they offer several advantages in terms of performance and functionality.

In order to use NumPy, you first need to import the numpy module. Once imported, you can create a NumPy array by passing a Python list or tuple to the numpy.array() function.

NOTE: Numpy can create ranges with all statistical distribution, but we only want this:
=> discrete uniform distribution: all numbers have the same probability of showing up - this means that with an array with 1, 2, 3, the numbers have the same probability of showing up

In [5]:
import numpy as np #do this only once

Exercise 10 a) Create a **1D** numpy array using manual numeration.

In [ ]:
a = np.array([1,2,3,4])
a

array([1, 2, 3, 4])

Exercise 10 b) using function `range`.

(if numbers are consecutive we can use this function)

In [ ]:
a = np.array(range(1,4+1))
a

array([1, 2, 3, 4])

NOTE: in the range function with interval the firts number is always included, i.e., will show up, but the second number, i.e., the last of the interval is not included => so we need to put the number we want plus 1. in this case we want 4, so 5 (4+1)

Exercise 10 c) using function `np.arange`.

`np.arange()` is a NumPy function that creates an array of evenly spaced values within a specified range. It is similar to the built-in range() function in Python, but it returns a NumPy array instead of a Python list.

In [ ]:
a = np.arange(1,4+1)
a

array([1, 2, 3, 4])

Exercise 11 a) Create a **2D** numpy array suing manual numeration.

In [ ]:
a = np.array([[1, 2, 3, 4],
              [8, 9, 6, 5]])

Exercise 11 b) using manual function np.reshape.

In [ ]:
a = a.reshape(2,4)
a

array([[1, 2, 3, 4],
       [8, 9, 6, 5]])

Exercise 12 create a 2D numpy array (3 x 4), whose elements are randomly drawn from the discrete uniform distribution between 2 (inclusive) and 30 (inclusive)

In [ ]:
np.random.seed(7548)
a = np.random.randint(2, 30+1, size=3*4)
a = a.reshape(3,4)
a

array([[10,  7, 20, 15],
       [21, 10, 10,  8],
       [13, 20, 10, 19]])

NOTE: the reshape function allow us to present the array as a matrix.

Exercise 12 a) numpy array that contains the sum of each row of the numpy array.

TIP => row is represented by axis=1

In [ ]:
a.sum(axis=1)

array([52, 49, 62])

Exercise 12 b) numpy array that contains the sums of each column of the numpy array.

TIP => columns are represented by axis=0

In [ ]:
a.sum(axis=0)

array([44, 37, 40, 42])

What about the mean of the rows or colums?

In [ ]:
a.mean(axis=1)

array([13.  , 12.25, 15.5 ])

In [ ]:
a.mean(axis=0)

array([14.66666667, 12.33333333, 13.33333333, 14.        ])

# CURIOSITY
In Python, in sums, true becomes 1 and false becomes 0.

In [ ]:
np.sum(True)

1

In [ ]:
np.sum(False)

0

Exercise 12 c) numpy array that per row counts the number of elements greater than 15.

In [ ]:
np.sum(a > 15, axis=1)

array([1, 1, 2])

Exercise 12 d) numpy array that per column counts the number of elements lower than 20.

In [ ]:
np.sum(a < 20, axis=0)

array([2, 2, 2, 3])

Exercise 13


In [2]:
names = ['Mary', 'John', 'Catherine']
course1 = [15, 18, 19]
course2 = [18, 19, 14]
course3 = [14, 16, 17]
credits = [6, 4.5, 3] # of the above courses, respectively

Let's put the courses in the same array.

In [6]:
np.array([course1, course2, course3])

array([[15, 18, 19],
       [18, 19, 14],
       [14, 16, 17]])

NOTE: this array is not the same as have a table (a ordem dos dados) because the first row corresponds to the 1st column in the table (1st column is course1) So we need to transpose the marks (.T)


In [7]:
marks = np.array([course1, course2, course3])
marks = marks.T
marks

array([[15, 18, 14],
       [18, 19, 16],
       [19, 14, 17]])

Linhas => Notas da Maria em cada curso
Colunas => Notas dos três em cada curso

In [ ]:
num_rows = len(marks)
num_cols = len(marks[0])

print("Number of rows:", num_rows)
print("Number of columns:", num_cols)

Number of rows: 3
Number of columns: 3


NOTE: agora os valores do course 1 estão presentes na primeira coluna.

Let's multiply each course by its credits (course1*credits1), being credits 1=6

NOTE: notas do course1 estão na primeira coluna que é 3x3 e matriz dos credits é 3x1

In [8]:
credits = np.array([6, 4.5, 3])

marks*credits

array([[ 90. ,  81. ,  42. ],
       [108. ,  85.5,  48. ],
       [114. ,  63. ,  51. ]])

Linhas => credits*marks de cada um dos alunos
Colunas => credits*marks dos alunos em cada um dos cursos

Let's do the average f each student (rows).

In [9]:
np.sum(marks * credits, axis=1)/ np.sum(credits)

array([15.77777778, 17.88888889, 16.88888889])

Let's round these values to 2 decimal digits.

In [10]:
np.round(np.sum(marks * credits, axis=1)/ np.sum(credits),2)

array([15.78, 17.89, 16.89])

Let's do the average of each course (colums) axis=0

In [11]:
np.sum(marks * credits, axis=0) / np.sum(credits)

array([23.11111111, 17.        , 10.44444444])

Let's round these values in 2 decimal digits

In [12]:
np.round(np.sum(marks * credits, axis=0) / np.sum(credits),2)

array([23.11, 17.  , 10.44])

Exercise 14  create a 3D numpy array (4x2x5) whose elements are randomly drawn from the integer uniform distribution between 2 (inclusive) and 30 (exclusive)

Use np.random.seed(124)

3D numpy arrays are necessary for image processing for example (we will not use this application)

In [ ]:
np.random.seed(124)

a = np.random.randint(2, 30+1, size=4*2*5)
a = a.reshape(4,2,5)
a


array([[[16, 30, 19,  9,  4],
        [11,  2, 22, 22, 30]],

       [[16, 23, 23, 24, 30],
        [29, 11, 24, 16, 18]],

       [[10, 11, 11,  2,  4],
        [20, 19,  3, 11, 15]],

       [[13,  3, 27, 12,  6],
        [28,  3, 21, 23, 24]]])

Exercise 14 a) looking at the random numpy array we created what instruction leads to this numpy array

array([[55, 67, 80, 47, 44], 
       [88, 35, 70, 72, 87]])

É possível verificar que para obter o primeiro número da primeira linha temos que somar todos os 1os números da primeira linha tbm e assim consecutivamente.
Ou seja, o que estamos a fazer é somar os valores por coluna, então temos que usar axis=0.


In [ ]:
a.sum(axis=0)

array([[55, 67, 80, 47, 44],
       [88, 35, 70, 72, 87]])

Exercise 14 b) to create the following numpy array what instruction we need to use?

array([[27, 32, 41, 31, 34],
       [45, 34, 47, 40, 48],
       [30, 30, 14, 13, 19],
       [41,  6, 48, 35, 30]])

Aqui estamos a somar o primeiro elemento da primeira linha com o primeiro elemento da segunda linhas, então temos de usar axis=1

In [ ]:
a.sum(axis=1)

array([[27, 32, 41, 31, 34],
       [45, 34, 47, 40, 48],
       [30, 30, 14, 13, 19],
       [41,  6, 48, 35, 30]])

Exercise 13 c) to create the following numpy array what instruction we need to use?

array([[ 78,  87],
       [116,  98],
       [ 38,  68],
       [ 61,  99]])

O que estamos a fazer é somar todos os elementos de cada linha, então temos de usar axis=2



In [ ]:
a.sum(axis=2)


array([[ 78,  87],
       [116,  98],
       [ 38,  68],
       [ 61,  99]])

Exercise 15 very similar to exercise 14

In [ ]:
np.random.seed(212)

a = np.random.randint(2, 30+1, size=2*4*3)
a = a.reshape(2,4,3)
a

array([[[26,  4,  6],
        [11, 16,  3],
        [27, 14, 21],
        [17, 18, 17]],

       [[15, 18, 14],
        [ 3, 12, 24],
        [28, 19,  5],
        [14,  2,  7]]])

Somatório de todos os primeiros elementos de cada linha.

In [ ]:
a.sum(axis=1)

array([[81, 52, 47],
       [60, 51, 50]])

Somatório de todos os elementos de cada linha.

In [ ]:
a.sum(axis=2)

array([[36, 30, 62, 52],
       [47, 39, 52, 23]])

Somatório do primeiro elemento de cada linha de cada matriz com o primeiro elemento da primeira linha com a segunda matriz.

In [ ]:
a.sum(axis=0)

array([[41, 22, 20],
       [14, 28, 27],
       [55, 33, 26],
       [31, 20, 24]])

# PANDA DATAFRAMES

Pandas is a popular open-source data analysis library for Python. It provides data structures for efficient data manipulation and analysis. One of the key data structures in pandas is the DataFrame, which is a two-dimensional table-like data structure.

A DataFrame is a collection of rows and columns, similar to a spreadsheet or SQL table. It has an index that labels the rows and a set of columns that label the columns. Each column in a DataFrame can have a different data type, such as integers, floating-point numbers, and strings.

Exercise 16 a) program that creates a dataframe, `df`, using a dictionary.

In [ ]:
import pandas as pd

df = pd.DataFrame({'Students': ['John', 'Catherine', 'Mary'],
                   'Ages': [18,30,25]})

df

Students  Ages
0       John    18
1  Catherine    30
2       Mary    25

Exercise 16 b) using a 2D numpy array

In [ ]:
a = np.array([df['Students'].values, df['Ages'].values])
a

array([['John', 'Catherine', 'Mary'],
       [18, 30, 25]], dtype=object)

Transpose the array to make the students and ages as rows.

In [ ]:
a = a.T

print(a)

[['John' 18]
 ['Catherine' 30]
 ['Mary' 25]]


## READING AND WRITING DATA FILES

Exercise 17

In [ ]:
df_carradios = pd.read_excel('/content/data_carradios.xlsx')
df_carradios

perc_defec      bdate  team  training      datep  prizeq  prized
0          0.00 1981-07-02     8         1 2021-07-01       0       0
1         36.32 1992-06-14     6         0 2021-07-02       0       0
2         48.91 2003-05-28     7         0 2021-07-05     500       0
3         20.36 1992-06-14    10         0 2021-07-06       0       0
4         42.07 2003-05-28     7         0 2021-07-07     500       0
..          ...        ...   ...       ...        ...     ...     ...
995       21.50 1981-07-02     1         0 2021-12-13       0       0
996        7.53 1981-07-02     2         0 2021-12-14     500     600
997       27.70 1997-12-05    10         0 2021-12-15       0       0
998        0.00 1986-12-23     5         0 2021-12-16       0     600
999       38.39 1986-12-23     6         0 2021-12-17       0       0

[1000 rows x 7 columns]

Exercise 18

In [ ]:
df_marks = pd.read_csv('/content/course_marks.csv')
df_marks

Names  Test 1  Test 2  Test 3
0           Troy Day   12.43   14.13   14.65
1   Daniel Hernandez   13.22   14.47   11.95
2    Michelle Wilson   12.82   16.11   14.95
3         Sarah Moon   13.68   15.36   16.61
4       Devin Carter   13.28   15.77   15.46
5       John Griffin   14.97   16.13   14.81
6         Maria Gray   12.89   19.02   14.06
7    William Mcclure   14.06   16.47   14.32
8       Darrell King   10.21   15.13   14.20
9     Valerie Holder   15.58   15.38   16.66
10   Brandon Mullins   14.17   17.35   14.97
11   Makayla Whitney   15.86   19.20   18.45
12     David Higgins   13.73   15.41   13.81
13       Samuel Vega   13.39   18.90   15.70
14     Melissa Berry   17.61   18.39   16.60
15   Brandon Johnson   13.36   17.44   17.64
16        Ryan Weber   15.58   19.56   17.17
17     Karen Stanley   13.12   17.72   14.89
18       Ryan Spence   14.41   16.79   15.89
19     Ryan Harrison   13.10   17.23   15.67

Exercise 19 Write a program that writes the pandas dataframe df created in Question 16 to csv file.

Index = True appears like this:

  name  age gender

0  Bob   25   male

1  Amy   30 female

2  Tim   35   male

index = False, exclude the index 0, 1 and 2

also, we can set the index we want, for example, 'name'

`df = df.set_index('name')`

it appears like this:

name  age gender
           
Bob    25   male

Amy    30 female

Tim    35   male

In [ ]:
df.to_csv('mydf.csv', index=False)

Exercise 20

In [ ]:
df.to_excel('mydf.xlsx', index=False)

## FUNCTIONS

```
def name_of_function(parameters):
  line of code
  line of code
  line of code
  ...
  return result
```

DON'T FORGET THE : IN THE FUNCTIONS

Exercise 21 Consider the dataset course_marks.csv. The three tests have the weights 25%, 40% and 35%, respectively.

Exercise 21 a) Write a function that adds a column, named Final, to the dataframe with the final marks for each student.
ou seja, 15* 0.25+18* 0.4+12*0.35

In [ ]:
##Creating the function
def calc_score(t1, t2, t3):
  score = t1*0.25 + t2*0.4 + t3*0.35
  return score

Giving values to the function...

In [ ]:
#MAKING AN EXAMPLE
calc_score(15,18,12)

15.149999999999999

In [ ]:
#ADDING THE COLUMN FINAL
df_marks['Score'] = [calc_score(x[0], x[1], x[2]) for x in zip(df_marks['Test 1'], df_marks['Test 2'], df_marks['Test 3'])]
df_marks                            

Names  Test 1  Test 2  Test 3    Score
0           Troy Day   12.43   14.13   14.65  13.8870
1   Daniel Hernandez   13.22   14.47   11.95  13.2755
2    Michelle Wilson   12.82   16.11   14.95  14.8815
3         Sarah Moon   13.68   15.36   16.61  15.3775
4       Devin Carter   13.28   15.77   15.46  15.0390
5       John Griffin   14.97   16.13   14.81  15.3780
6         Maria Gray   12.89   19.02   14.06  15.7515
7    William Mcclure   14.06   16.47   14.32  15.1150
8       Darrell King   10.21   15.13   14.20  13.5745
9     Valerie Holder   15.58   15.38   16.66  15.8780
10   Brandon Mullins   14.17   17.35   14.97  15.7220
11   Makayla Whitney   15.86   19.20   18.45  18.1025
12     David Higgins   13.73   15.41   13.81  14.4300
13       Samuel Vega   13.39   18.90   15.70  16.4025
14     Melissa Berry   17.61   18.39   16.60  17.5685
15   Brandon Johnson   13.36   17.44   17.64  16.4900
16        Ryan Weber   15.58   19.56   17.17  17.7285
17     Karen Stanley   13.12   17.72   14.89  15.5795
18       Ryan Spence   14.41   16.79   15.89  15.8800
19     Ryan Harrison   13.10   17.23   15.67  15.6515

Exercise 21 b) add the same column using pandas only

In [ ]:
df_marks['Score2'] = df_marks['Test 1'].mul(0.25)+df_marks['Test 2'].mul(0.4)+df_marks['Test 3'].mul(0.35)
df_marks

Names  Test 1  Test 2  Test 3    Score   Score2
0           Troy Day   12.43   14.13   14.65  13.8870  13.8870
1   Daniel Hernandez   13.22   14.47   11.95  13.2755  13.2755
2    Michelle Wilson   12.82   16.11   14.95  14.8815  14.8815
3         Sarah Moon   13.68   15.36   16.61  15.3775  15.3775
4       Devin Carter   13.28   15.77   15.46  15.0390  15.0390
5       John Griffin   14.97   16.13   14.81  15.3780  15.3780
6         Maria Gray   12.89   19.02   14.06  15.7515  15.7515
7    William Mcclure   14.06   16.47   14.32  15.1150  15.1150
8       Darrell King   10.21   15.13   14.20  13.5745  13.5745
9     Valerie Holder   15.58   15.38   16.66  15.8780  15.8780
10   Brandon Mullins   14.17   17.35   14.97  15.7220  15.7220
11   Makayla Whitney   15.86   19.20   18.45  18.1025  18.1025
12     David Higgins   13.73   15.41   13.81  14.4300  14.4300
13       Samuel Vega   13.39   18.90   15.70  16.4025  16.4025
14     Melissa Berry   17.61   18.39   16.60  17.5685  17.5685
15   Brandon Johnson   13.36   17.44   17.64  16.4900  16.4900
16        Ryan Weber   15.58   19.56   17.17  17.7285  17.7285
17     Karen Stanley   13.12   17.72   14.89  15.5795  15.5795
18       Ryan Spence   14.41   16.79   15.89  15.8800  15.8800
19     Ryan Harrison   13.10   17.23   15.67  15.6515  15.6515

Exercise 21 c) add the same column, using numpy only.

In [ ]:
import numpy as np
import pandas as pd

In [ ]:
df_marks2 = pd.read_csv('/content/course_marks.csv')
df_marks2

Names  Test 1  Test 2  Test 3
0           Troy Day   12.43   14.13   14.65
1   Daniel Hernandez   13.22   14.47   11.95
2    Michelle Wilson   12.82   16.11   14.95
3         Sarah Moon   13.68   15.36   16.61
4       Devin Carter   13.28   15.77   15.46
5       John Griffin   14.97   16.13   14.81
6         Maria Gray   12.89   19.02   14.06
7    William Mcclure   14.06   16.47   14.32
8       Darrell King   10.21   15.13   14.20
9     Valerie Holder   15.58   15.38   16.66
10   Brandon Mullins   14.17   17.35   14.97
11   Makayla Whitney   15.86   19.20   18.45
12     David Higgins   13.73   15.41   13.81
13       Samuel Vega   13.39   18.90   15.70
14     Melissa Berry   17.61   18.39   16.60
15   Brandon Johnson   13.36   17.44   17.64
16        Ryan Weber   15.58   19.56   17.17
17     Karen Stanley   13.12   17.72   14.89
18       Ryan Spence   14.41   16.79   15.89
19     Ryan Harrison   13.10   17.23   15.67

Transform the dataframe into a numpy array

In [ ]:
df_marks2.values


array([['Troy Day', 12.43, 14.13, 14.65],
       ['Daniel Hernandez', 13.22, 14.47, 11.95],
       ['Michelle Wilson', 12.82, 16.11, 14.95],
       ['Sarah Moon', 13.68, 15.36, 16.61],
       ['Devin Carter', 13.28, 15.77, 15.46],
       ['John Griffin', 14.97, 16.13, 14.81],
       ['Maria Gray', 12.89, 19.02, 14.06],
       ['William Mcclure', 14.06, 16.47, 14.32],
       ['Darrell King', 10.21, 15.13, 14.2],
       ['Valerie Holder', 15.58, 15.38, 16.66],
       ['Brandon Mullins', 14.17, 17.35, 14.97],
       ['Makayla Whitney', 15.86, 19.2, 18.45],
       ['David Higgins', 13.73, 15.41, 13.81],
       ['Samuel Vega', 13.39, 18.9, 15.7],
       ['Melissa Berry', 17.61, 18.39, 16.6],
       ['Brandon Johnson', 13.36, 17.44, 17.64],
       ['Ryan Weber', 15.58, 19.56, 17.17],
       ['Karen Stanley', 13.12, 17.72, 14.89],
       ['Ryan Spence', 14.41, 16.79, 15.89],
       ['Ryan Harrison', 13.1, 17.23, 15.67]], dtype=object)

: => means every rows

1: => the column of index 1 and all columns on the right

df.values[:, 1:] (df.values means numpy array and [] means all rows and the column of index 1 and all columns on the right.

Multiplicar as notas de cada um dos três testes pelo peso de cada um...

In [ ]:
df_marks2.values[:, 1:] * np.array([0.25, 0.4, 0.35])

array([[3.1075, 5.652000000000001, 5.1274999999999995],
       [3.305, 5.788, 4.182499999999999],
       [3.205, 6.444, 5.232499999999999],
       [3.42, 6.144, 5.8134999999999994],
       [3.32, 6.308, 5.411],
       [3.7425, 6.452, 5.1834999999999996],
       [3.2225, 7.6080000000000005, 4.921],
       [3.515, 6.588, 5.012],
       [2.5525, 6.0520000000000005, 4.97],
       [3.895, 6.152000000000001, 5.8309999999999995],
       [3.5425, 6.940000000000001, 5.2395],
       [3.965, 7.68, 6.4575],
       [3.4325, 6.164000000000001, 4.8335],
       [3.3475, 7.56, 5.494999999999999],
       [4.4025, 7.356000000000001, 5.8100000000000005],
       [3.34, 6.976000000000001, 6.1739999999999995],
       [3.895, 7.824, 6.0095],
       [3.28, 7.088, 5.2115],
       [3.6025, 6.716, 5.5615],
       [3.275, 6.892, 5.4845]], dtype=object)

Somar os três valores de cada linha para termos a nota final...

In [ ]:
np.sum(df_marks2.values[:, 1:] * np.array([0.25, 0.4, 0.35]), axis=1)

array([13.887, 13.2755, 14.881499999999999, 15.3775, 15.039, 15.378,
       15.7515, 15.114999999999998, 13.5745, 15.878, 15.722000000000001,
       18.1025, 14.43, 16.402499999999996, 17.5685, 16.490000000000002,
       17.7285, 15.5795, 15.879999999999999, 15.651499999999999],
      dtype=object)

Turn Final3 into a dataframe.

In [ ]:
df_marks2['Final3'] = np.sum(df_marks2.values[:, 1:] * np.array([0.25, 0.4, 0.35]), axis=1)
df_marks2


Names  Test 1  Test 2  Test 3   Final3
0           Troy Day   12.43   14.13   14.65   13.887
1   Daniel Hernandez   13.22   14.47   11.95  13.2755
2    Michelle Wilson   12.82   16.11   14.95  14.8815
3         Sarah Moon   13.68   15.36   16.61  15.3775
4       Devin Carter   13.28   15.77   15.46   15.039
5       John Griffin   14.97   16.13   14.81   15.378
6         Maria Gray   12.89   19.02   14.06  15.7515
7    William Mcclure   14.06   16.47   14.32   15.115
8       Darrell King   10.21   15.13   14.20  13.5745
9     Valerie Holder   15.58   15.38   16.66   15.878
10   Brandon Mullins   14.17   17.35   14.97   15.722
11   Makayla Whitney   15.86   19.20   18.45  18.1025
12     David Higgins   13.73   15.41   13.81    14.43
13       Samuel Vega   13.39   18.90   15.70  16.4025
14     Melissa Berry   17.61   18.39   16.60  17.5685
15   Brandon Johnson   13.36   17.44   17.64    16.49
16        Ryan Weber   15.58   19.56   17.17  17.7285
17     Karen Stanley   13.12   17.72   14.89  15.5795
18       Ryan Spence   14.41   16.79   15.89    15.88
19     Ryan Harrison   13.10   17.23   15.67  15.6515

# DIFFERENT WAYS TO SELECT THE COLUMNS WE WANT

Another way is to use pandas to select the columns we need.

df.iloc[:,1:] <= this only works with numerical indexing.

df.loc[:, [ ] ] <=  [ ] means list of columns <= it works with columns


In [ ]:
df_marks2.iloc[:, 1:]

Test 1  Test 2  Test 3   Final3
0    12.43   14.13   14.65   13.887
1    13.22   14.47   11.95  13.2755
2    12.82   16.11   14.95  14.8815
3    13.68   15.36   16.61  15.3775
4    13.28   15.77   15.46   15.039
5    14.97   16.13   14.81   15.378
6    12.89   19.02   14.06  15.7515
7    14.06   16.47   14.32   15.115
8    10.21   15.13   14.20  13.5745
9    15.58   15.38   16.66   15.878
10   14.17   17.35   14.97   15.722
11   15.86   19.20   18.45  18.1025
12   13.73   15.41   13.81    14.43
13   13.39   18.90   15.70  16.4025
14   17.61   18.39   16.60  17.5685
15   13.36   17.44   17.64    16.49
16   15.58   19.56   17.17  17.7285
17   13.12   17.72   14.89  15.5795
18   14.41   16.79   15.89    15.88
19   13.10   17.23   15.67  15.6515

In values...

In [ ]:
df_marks2.iloc[:, 1:].values

array([[12.43, 14.13, 14.65, 13.887],
       [13.22, 14.47, 11.95, 13.2755],
       [12.82, 16.11, 14.95, 14.881499999999999],
       [13.68, 15.36, 16.61, 15.3775],
       [13.28, 15.77, 15.46, 15.039],
       [14.97, 16.13, 14.81, 15.378],
       [12.89, 19.02, 14.06, 15.7515],
       [14.06, 16.47, 14.32, 15.114999999999998],
       [10.21, 15.13, 14.2, 13.5745],
       [15.58, 15.38, 16.66, 15.878],
       [14.17, 17.35, 14.97, 15.722000000000001],
       [15.86, 19.2, 18.45, 18.1025],
       [13.73, 15.41, 13.81, 14.43],
       [13.39, 18.9, 15.7, 16.402499999999996],
       [17.61, 18.39, 16.6, 17.5685],
       [13.36, 17.44, 17.64, 16.490000000000002],
       [15.58, 19.56, 17.17, 17.7285],
       [13.12, 17.72, 14.89, 15.5795],
       [14.41, 16.79, 15.89, 15.879999999999999],
       [13.1, 17.23, 15.67, 15.651499999999999]], dtype=object)

Another way to do the same thing...

In [ ]:
df.loc[:, ['Test 1', 'Test 2', 'Test 3']]

Test 1  Test 2  Test 3
0    12.43   14.13   14.65
1    13.22   14.47   11.95
2    12.82   16.11   14.95
3    13.68   15.36   16.61
4    13.28   15.77   15.46
5    14.97   16.13   14.81
6    12.89   19.02   14.06
7    14.06   16.47   14.32
8    10.21   15.13   14.20
9    15.58   15.38   16.66
10   14.17   17.35   14.97
11   15.86   19.20   18.45
12   13.73   15.41   13.81
13   13.39   18.90   15.70
14   17.61   18.39   16.60
15   13.36   17.44   17.64
16   15.58   19.56   17.17
17   13.12   17.72   14.89
18   14.41   16.79   15.89
19   13.10   17.23   15.67

Exercise 22 Consider the dataset warehouse_sales.csv



In [ ]:
df_sales = pd.read_csv('/content/warehouse_sales.csv')
df_sales

Product   Price        Date  Vat Rate
0     For repair Cotton Pizza  356.00  2022-12-30      0.20
1               Plastic Chair  909.56  2022-12-01      0.15
2                       Table  520.65  2022-10-23      0.20
3               Plastic Salad  277.53  2022-11-01      0.20
4                        Fish   89.94  2022-10-21      0.05
...                       ...     ...         ...       ...
9995                Soft Bike  311.18  2022-11-10      0.10
9996              Wooden Tuna  726.36  2022-10-21      0.10
9997   Practical Wooden Pizza  749.95  2023-01-19      0.15
9998          Rubber Sausages  674.20  2022-11-05      0.20
9999          Wooden Sausages  508.32  2023-01-08      0.05

[10000 rows x 4 columns]

a) write a function that adds a column named Total to the dataframe, which contains for each sale the total amount to pay including VAT

In [ ]:
def total_amount(p, vat):
  amount = p+p*vat
  return amount

In [ ]:
total_amount(5, 5)

30

In [ ]:
df_sales['Total'] = [total_amount(x[0], x[1]) for x in zip(df_sales['Price'], df_sales['Vat Rate'])]
df_sales

Product   Price        Date  Vat Rate      Total
0     For repair Cotton Pizza  356.00  2022-12-30      0.20   427.2000
1               Plastic Chair  909.56  2022-12-01      0.15  1045.9940
2                       Table  520.65  2022-10-23      0.20   624.7800
3               Plastic Salad  277.53  2022-11-01      0.20   333.0360
4                        Fish   89.94  2022-10-21      0.05    94.4370
...                       ...     ...         ...       ...        ...
9995                Soft Bike  311.18  2022-11-10      0.10   342.2980
9996              Wooden Tuna  726.36  2022-10-21      0.10   798.9960
9997   Practical Wooden Pizza  749.95  2023-01-19      0.15   862.4425
9998          Rubber Sausages  674.20  2022-11-05      0.20   809.0400
9999          Wooden Sausages  508.32  2023-01-08      0.05   533.7360

[10000 rows x 5 columns]

Exercise 22 b) add the same column using pandas only

In [ ]:
df_sales['Total2'] = df_sales['Price']+df_sales['Price'].mul(df_sales['Vat Rate'])
df_sales

Product   Price        Date  Vat Rate      Total  \
0     For repair Cotton Pizza  356.00  2022-12-30      0.20   427.2000   
1               Plastic Chair  909.56  2022-12-01      0.15  1045.9940   
2                       Table  520.65  2022-10-23      0.20   624.7800   
3               Plastic Salad  277.53  2022-11-01      0.20   333.0360   
4                        Fish   89.94  2022-10-21      0.05    94.4370   
...                       ...     ...         ...       ...        ...   
9995                Soft Bike  311.18  2022-11-10      0.10   342.2980   
9996              Wooden Tuna  726.36  2022-10-21      0.10   798.9960   
9997   Practical Wooden Pizza  749.95  2023-01-19      0.15   862.4425   
9998          Rubber Sausages  674.20  2022-11-05      0.20   809.0400   
9999          Wooden Sausages  508.32  2023-01-08      0.05   533.7360   

         Total2  
0      427.2000  
1     1045.9940  
2      624.7800  
3      333.0360  
4       94.4370  
...         ...  
9995   342.2980  
9996   798.9960  
9997   862.4425  
9998   809.0400  
9999   533.7360  

[10000 rows x 6 columns]

Exercise 22 c) add the same column using numpy only

In [ ]:
df_sales.values

array([['For repair Cotton Pizza', 356.0, '2022-12-30', 0.2, 427.2,
        427.2],
       ['Plastic Chair', 909.56, '2022-12-01', 0.15, 1045.994, 1045.994],
       ['Table', 520.65, '2022-10-23', 0.2, 624.78, 624.78],
       ...,
       ['Practical Wooden Pizza', 749.95, '2023-01-19', 0.15,
        862.4425000000001, 862.4425000000001],
       ['Rubber Sausages', 674.2, '2022-11-05', 0.2, 809.0400000000001,
        809.0400000000001],
       ['Wooden Sausages', 508.32, '2023-01-08', 0.05, 533.736, 533.736]],
      dtype=object)

In [ ]:
df_sales.values[:,1:]

array([[356.0, '2022-12-30', 0.2, 427.2, 427.2],
       [909.56, '2022-12-01', 0.15, 1045.994, 1045.994],
       [520.65, '2022-10-23', 0.2, 624.78, 624.78],
       ...,
       [749.95, '2023-01-19', 0.15, 862.4425000000001, 862.4425000000001],
       [674.2, '2022-11-05', 0.2, 809.0400000000001, 809.0400000000001],
       [508.32, '2023-01-08', 0.05, 533.736, 533.736]], dtype=object)

In [ ]:
df_sales.loc[:, ['Price', 'Vat Rate']]

Price  Vat Rate
0     356.00      0.20
1     909.56      0.15
2     520.65      0.20
3     277.53      0.20
4      89.94      0.05
...      ...       ...
9995  311.18      0.10
9996  726.36      0.10
9997  749.95      0.15
9998  674.20      0.20
9999  508.32      0.05

[10000 rows x 2 columns]

In [ ]:
np.price = df_sales.loc[:, ['Price']]
np.vatRate = df_sales.loc[:, ['Vat Rate']]

In [ ]:
np.price.values

array([[356.  ],
       [909.56],
       [520.65],
       ...,
       [749.95],
       [674.2 ],
       [508.32]])

In [ ]:
np.vatRate.values

array([[0.2 ],
       [0.15],
       [0.2 ],
       ...,
       [0.15],
       [0.2 ],
       [0.05]])

In [ ]:
np.price.values+np.price.values*np.vatRate.values

array([[ 427.2   ],
       [1045.994 ],
       [ 624.78  ],
       ...,
       [ 862.4425],
       [ 809.04  ],
       [ 533.736 ]])

In [ ]:
df_sales['Total3'] = np.price.values+np.price.values*np.vatRate.values
df_sales


Product   Price        Date  Vat Rate      Total  \
0     For repair Cotton Pizza  356.00  2022-12-30      0.20   427.2000   
1               Plastic Chair  909.56  2022-12-01      0.15  1045.9940   
2                       Table  520.65  2022-10-23      0.20   624.7800   
3               Plastic Salad  277.53  2022-11-01      0.20   333.0360   
4                        Fish   89.94  2022-10-21      0.05    94.4370   
...                       ...     ...         ...       ...        ...   
9995                Soft Bike  311.18  2022-11-10      0.10   342.2980   
9996              Wooden Tuna  726.36  2022-10-21      0.10   798.9960   
9997   Practical Wooden Pizza  749.95  2023-01-19      0.15   862.4425   
9998          Rubber Sausages  674.20  2022-11-05      0.20   809.0400   
9999          Wooden Sausages  508.32  2023-01-08      0.05   533.7360   

         Total2     Total3  
0      427.2000   427.2000  
1     1045.9940  1045.9940  
2      624.7800   624.7800  
3      333.0360   333.0360  
4       94.4370    94.4370  
...         ...        ...  
9995   342.2980   342.2980  
9996   798.9960   798.9960  
9997   862.4425   862.4425  
9998   809.0400   809.0400  
9999   533.7360   533.7360  

[10000 rows x 7 columns]